### PR TITLE
Validate http headers names

### PR DESCRIPTION
**Pull request by vhvb1989**
As per Ahson request, we are adding validation for headers names.

For Http Requests, validation is done as soon as customer try to append one header to some http request. We return AZ_ERROR_ARG as the input args are not valid to append a header.

For Http Response, ~~I propose in this PR to add a policy before transport policy that will handle validation after response comes back from transport policy. This way, if customer wants to remove this, they can remove this policy from the pipeline and trust that Service will never return a bad header back (this would provide some speed because validating response headers needs to parse the response and loop through all headers in the response)~~
 Http Response won't validate headers until customer call az_http_response_get_next_header()

### CODE REVIEW

**Comment by JeffreyRichter** [comment]
For request headers, I think we can check them with a precondition; what would a customer do with AZ_ERROR_ARG?

For response headers, the policy idea is an interesting one; I never thought of that myself. The downside to this is that you're ultimately scanning over the data twice: once in the policy & once again as we return headers to the customer. We always have to scan the response to return headers to the customer. So, while scanning, it might just be faster to check the characters against valid characters. We could make this an http_response option set when initializing it IF the perf hit is too great.

File: sdk/core/core/src/az_http_header_validation_private.h [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  'w', // 119 ->  w
+  'x', // 120 ->  x
+  'y', // 121 ->  y
+  'z', // 122 ->  z
+  0, // 123 ->  {
+  '|', // 124 ->  |
+  0, // 125 ->  }
+  '~', // 126 ->  ~
+  0 // 127 ->  DEL
+};
```
Comment (JeffreyRichter) [inline sdk/core/core/src/az_http_header_validation_private.h:148]: Add a comment about this that 128-255 is all zeros (not valid) characters

File: sdk/core/core/src/az_http_header_validation_private.h [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+// Loop all characters from name and validate each.
+AZ_NODISCARD AZ_INLINE bool az_http_is_valid_header_name(az_span name)
+{
+  uint8_t* name_ptr = az_span_ptr(name);
+  for (int32_t i = 0; i < az_span_size(name); i++)
+  {
+    uint8_t c = name_ptr[i];
+    if (c > 127 || az_http_valid_token[c] == 0)
```
Comment (JeffreyRichter) [inline sdk/core/core/src/az_http_header_validation_private.h:157]: Wait: If you have the >127 check then the array can be 128 size, not 256.
But, this test makes the check slower.

**Comment by vhvb1989** [comment]
> For request headers, I think we can check them with a precondition; what would a customer do with AZ_ERROR_ARG?

The only problem whit this is if customer won't set header statically and will get the values for header name and value from I/O. Anyway, I have changed it to PRECONDITION and eventually we will see if this becomes a problem in the future.  Worst case will be customer adding a header that Service will validate and return INVALID HTTP Request back.

> For response headers, the policy idea is an interesting one; I never thought of that myself. The downside to this is that you're ultimately scanning over the data twice: once in the policy & once again as we return headers to the customer. We always have to scan the response to return headers to the customer. So, while scanning, it might just be faster to check the characters against valid characters. We could make this an http_response option set when initializing it IF the perf hit is too great.

The problem of waiting to validate header until we are parsing the response for the user is that we won't fail the request. We will return AZ_OK and fill the az_http_response with the invalid data.
Then during parsing, we will return the error.
My understanding is that we want to fail the request on this case. We would be still returning all the response to customer, but he will know that it's not valid in advance.

If a Service like KeyVault or Storage would review this SDK client and decide they won't never return an invalid header, they can update the SDK client and remove the policy.

By adding an option for az_http_response, it would be final customer decision if they want to have this validation or not.

File: sdk/core/core/src/az_http_header_validation_private.h [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+#include <az_span.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <_az_cfg_prefix.h>
+
+// Bit array from ASCII to valid. Every zero in array means invalid character, while non-zeros
+// return the valid character.
+static const uint8_t az_http_valid_token[256] = {
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_http_header_validation_private.h:24]: nit: consider representing this in a compact 16x16 grid where we just have 0/1 binary values rather than 0 and non-zero.

File: sdk/core/core/src/az_http_header_validation_private.h [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  'w', // 119 ->  w
+  'x', // 120 ->  x
+  'y', // 121 ->  y
+  'z', // 122 ->  z
+  0, // 123 ->  {
+  '|', // 124 ->  |
+  0, // 125 ->  }
+  '~', // 126 ->  ~
+  0 // 127 ->  DEL
+  // ...128-255 is all zeros (not valid) characters
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_http_header_validation_private.h:153]: Are you sure everything else will be zero-initialized (for all supported compilers)?
cc @barcharcraz

File: sdk/core/core/test/cmocka/test_az_http.c [hunk 5]
```diff
@@ -357,6 +357,58 @@ enable_precondition_check_tests()
...
+        &hrb,
+        &az_context_app,
+        az_http_method_get(),
+        url_span,
+        az_span_size(url_span),
+        header_span,
+        AZ_SPAN_FROM_STR("body")));
+
+    assert_precondition_checked(az_http_request_append_header(
+        &hrb, AZ_SPAN_FROM_STR("(headerName)"), hrb_header_content_type_token));
```
Comment (ahsonkhan) [inline sdk/core/core/test/cmocka/test_az_http.c:381]: nit: add other asserts on the contents of the `header_span`/`hrb`

Same below.

File: sdk/core/core/src/az_http_policy.c [hunk 6]
```diff
@@ -77,6 +79,34 @@ AZ_NODISCARD az_result az_http_pipeline_policy_credential(
...
+  // Validate headers on response
+  // 1. parse status line to have access to headers
+  az_http_response_status_line status = { 0 };
+  AZ_RETURN_IF_FAILED(az_http_response_get_status_line(p_response, &status));
+
+  // loop headers and validate
+  for (az_pair header;
+       az_http_response_get_next_header(p_response, &header) != AZ_ERROR_ITEM_NOT_FOUND;)
+  {
+    if (!az_http_is_valid_header_name(header.key))
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_http_policy.c:101]: This causes double parsing, which is a bit unfortunate. Can we design it differently to avoid the validation overhead?

File: sdk/core/core/test/cmocka/test_az_policy.c [hunk 7]
```diff
@@ -97,6 +97,96 @@ void test_az_http_pipeline_policy_telemetry(void** state)
...
+            },
+        };
+  az_http_response response = { 0 };
+  assert_return_code(
+      az_http_response_init(
+          &response,
+          AZ_SPAN_FROM_STR("HTTP/1.1 404 Not Found\r\n"
+                           "Header11: Value11\r\n"
+                           "Header22: NNNNOOOOPPPPQQQQRRRRSSSSTTTTUUUUVVVVWWWWXXXXYYYYZZZZ\r\n"
+                           "Header33:\r\n"
```
Comment (ahsonkhan) [inline sdk/core/core/test/cmocka/test_az_policy.c:180]: Add test cases for invalid white space.

File: sdk/core/core/test/cmocka/test_az_policy.c [hunk 8]
```diff
@@ -97,6 +97,96 @@ void test_az_http_pipeline_policy_telemetry(void** state)
...
+                           "He@ader44: cba888888777777666666555555444444333333222222111111\r\n"
```
Comment (ahsonkhan) [inline sdk/core/core/test/cmocka/test_az_policy.c:181]: Also, value must be USASCII
https://tools.ietf.org/html/rfc7230#section-3.2
https://tools.ietf.org/html/rfc7230#section-1.2
> and VCHAR (any visible [USASCII] character).

https://www.fileformat.info/info/charset/US-ASCII/list.htm

File: sdk/core/core/src/az_http_header_validation_private.h (hunk 3 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_http_header_validation_private.h:24]: our auto-formatter didn't like 16x16.  it takes control over! :)
That's one of the motivation of having this as an individual header.  C-world people is used to long files, I'd say.

I used non-zero and the value instead in casa we want to reuse the same array to lower-Case or something like that.  There's not really any significant benefit from one of other

File: sdk/core/core/src/az_http_header_validation_private.h (hunk 4 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_http_header_validation_private.h:153]: will add a test for that 👍

File: sdk/core/core/test/cmocka/test_az_policy.c (hunk 7 above)
Comment (vhvb1989) [reply sdk/core/core/test/cmocka/test_az_policy.c:180]: already there

File: sdk/core/core/test/cmocka/test_az_policy.c (hunk 8 above)
Comment (vhvb1989) [reply sdk/core/core/test/cmocka/test_az_policy.c:181]: Can you elaborate on this comment a little further?  Its is just  a comment? or what is it asking for?

File: sdk/core/core/src/az_http_policy.c (hunk 6 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_http_policy.c:101]: Well, it depends. 
Let me tag @JeffreyRichter  for this.

So, If we want to fail request if the response comes with an invalid header, and without making a double parsing, one option is to build a new API for httpResponse that transport-layer would use to write the response.
Right now we only provide an az_span to transport-layer and we delegate the responsibility of writing a valid HttpResponse to it. So, if there is any bug on any transport layer to write a valid RFC response (status lines, headers or even body).
Now, consider adding next API and hiding the az_span buffer from transport-layer
```c
az_http_response_write_status_line(httpResponse *r, az_span status_line);
az_http_response_append_header(httpResponse *r, az_span header);
az_http_response_append_chuck_body(httpResponse *r, az_span bodyChunk);
```
If we decide to add this, we would be taking control of the validation and correctness of any transport-layer. Then we can fail the request as soon as trasnport-layer gives us an invalid header.
Downside is that we need to force any transport-layer to always write status-line first, then headers and finally a body (to maintain the valid structure)

**Comment by JeffreyRichter** [comment]
No service will ever intentionally return invalid headers. So, the chance of discovering an invalid header is really zero. This checking is just for paranoia and so we don’t want to introduce infrastructure (like a policy) or significant CPU wastage (like scanning twice) to address something that should never happen.
It’s OK to just fail returning a header pair because this will never actually happen. Today, I would just check when returning a header and NOT offer a way for a customer to turn this off as it is an extra know that we have to document, test, support, explain, etc. IF we hear that perf is bad, then we can offer a switch to turn it off in a non-breaking way.


  *   Jeffrey Richter (watch Architecting Distributed Cloud Apps<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fazure.microsoft.com%2Ftraining%2Fdistributed-cloud-apps%2F&data=02%7C01%7Cjeffreyr%40microsoft.com%7Cc2139bebacc8433bd13308d76a014af0%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637094426089210301&sdata=GjJbhammA5C%2F1Pve%2FXjbMq8wjjpGr2W1KSrYaCmzq7E%3D&reserved=0>)

File: sdk/core/core/src/az_http_policy.c (hunk 6 above)
Comment (JeffreyRichter) [reply sdk/core/core/src/az_http_policy.c:101]: I don't think we need to invalidate the entire response.

File: sdk/core/core/test/cmocka/test_az_policy.c (hunk 8 above)
Comment (ahsonkhan) [reply sdk/core/core/test/cmocka/test_az_policy.c:181]: To be compliant, we need to validate that the header-value doesn't contain anything that isn't in the USASCII range. That means no byte value over 127.

File: sdk/core/core/test/cmocka/test_az_http.c [hunk 9]
```diff
@@ -357,6 +358,82 @@ enable_precondition_check_tests()
...
+    assert_memory_equal(header_buf, headers._internal.ptr, size);
+  }
+}
+
+static void test_http_request_header_validation_range(void** state)
+{
+  (void)state;
+  {
+    // just make sure compiler will set anything greater than 127 to zero
+    for (uint8_t value = 127; value > 127; value++) // break on unsigned back to zero
```
Comment (ahsonkhan) [inline sdk/core/core/test/cmocka/test_az_http.c:430]: Why rely on overflowing?
Change the boundary condition to be 256.
```suggestion
    for (uint8_t value = 127; value < 256; value++)
```

File: sdk/core/core/test/cmocka/test_az_policy.c (hunk 8 above)
Comment (ahsonkhan) [reply sdk/core/core/test/cmocka/test_az_policy.c:181]: On second thought, that might be too expensive. Up to you.

File: sdk/core/core/test/cmocka/test_az_http.c (hunk 9 above)
Comment (vhvb1989) [reply sdk/core/core/test/cmocka/test_az_http.c:430]: using `< 256` is an infinite loop.  Compiler throws a 'always true condition` con compilation time.
uint8_t range is 0-255.

File: sdk/core/core/test/cmocka/test_az_policy.c (hunk 8 above)
Comment (vhvb1989) [reply sdk/core/core/test/cmocka/test_az_policy.c:181]: won't add validation for header value

File: sdk/core/core/test/cmocka/test_az_http.c (hunk 9 above)
Comment (ahsonkhan) [reply sdk/core/core/test/cmocka/test_az_http.c:430]: Oh, lol, right. Good point. Use `int32_t` then and cast?

File: sdk/core/core/test/cmocka/test_az_policy.c (hunk 7 above)
Comment (ahsonkhan) [reply sdk/core/core/test/cmocka/test_az_policy.c:180]: Can you point me to the test? I couldn't find a test where we have a response with whitespace  between the header and colon.

File: sdk/core/core/test/cmocka/test_az_http.c (hunk 5 above)
Comment (vhvb1989) [reply sdk/core/core/test/cmocka/test_az_http.c:382]: done

File: sdk/core/core/test/cmocka/test_az_http.c (hunk 9 above)
Comment (vhvb1989) [reply sdk/core/core/test/cmocka/test_az_http.c:430]: no need to cast if we can directly use uint8_t.  And this is at the end, just a test, so, would be fine either way. 
I would just keep it with no casting

File: sdk/core/core/test/cmocka/test_az_policy.c (hunk 7 above)
Comment (vhvb1989) [reply sdk/core/core/test/cmocka/test_az_policy.c]: that's part of the trimming.  He don't fail if there is spaces before or after colon.
We trim the name and value for customer instead

File: sdk/core/core/inc/az_result.h [hunk 10]
```diff
@@ -128,6 +128,8 @@ typedef enum
   AZ_ERROR_HTTP_RESPONSE_OVERFLOW = _az_RESULT_MAKE_ERROR(_az_FACILITY_HTTP, 5),
   AZ_ERROR_HTTP_RESPONSE_COULDNT_RESOLVE_HOST = _az_RESULT_MAKE_ERROR(_az_FACILITY_HTTP, 6),
 
+  AZ_ERROR_HTTP_RESPONSE_CONTAINS_INVALID_HEADERS = _az_RESULT_MAKE_ERROR(_az_FACILITY_HTTP, 7),
```
Comment (JeffreyRichter) [inline sdk/core/core/inc/az_result.h:131]: If this is returned when getting a single header, then it should be renamed to something like: AZ_ERROR_HTTP_CORRUPT_RESPONSE_HEADER

File: sdk/core/core/src/az_http_header_validation_private.h [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+};
+
+// Loop all characters from name and validate each.
+AZ_NODISCARD AZ_INLINE bool az_http_is_valid_header_name(az_span name)
```
Comment (JeffreyRichter) [inline sdk/core/core/src/az_http_header_validation_private.h:157]: Is there no way to do this test WHILE scanning the response buffer to get the span for the header?
This function requires scanning twice: once while scanning for the header value and then again to test the characters.

Comment (JeffreyRichter) [inline sdk/core/core/src/az_http_header_validation_private.h:157]: This is just for request headers, right?
If so, a comment about that would be good and if we only do this when preconditions are on, then maybe surround this with the #ifdef for precondition checks.

Comment (vhvb1989) [reply sdk/core/core/src/az_http_header_validation_private.h:157]: yes, only for http request were we get an az_span for header name.
Will put it inside #ifndef  👍

File: sdk/core/core/src/az_http_header_validation_private.h [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+  0, // 23 ->  end of trans. block
+  0, // 24 ->  cancel
+  0, // 25 ->  end of medium
+  0, // 26 ->  substitute
+  0, // 27 ->  escape
+  0, // 28 ->  file separator
+  0, // 29 ->  group separator
+  0, // 30 ->  record separator
+  0, // 31 ->  unit separator
+  ' ', // 32 ->  space
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_http_header_validation_private.h:57]: I don't think the space (' ') should be allowed within header names.

File: sdk/core/core/src/az_http_header_validation_private.h [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+  'q', // 81 ->  Q
+  'r', // 82 ->  R
+  's', // 83 ->  S
+  't', // 84 ->  T
+  'u', // 85 ->  U
+  'v', // 86 ->  V
+  'w', // 87 ->  W
+  'x', // 88 ->  X
+  'y', // 89 ->  Y
+  'z', // 90 ->  Z
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_http_header_validation_private.h:115]: These are all wrong and should be capitalized.

Luckily, since they are all non-zero, they are still considered valid within header names so things should continue to work.

Please add tests where you have a header name containing all valid/allowed characters and one where one of the characters is invalid (loop through each of the remaining invalid characters). We need to verify and make sure we only reject invalid header names and not valid ones.

File: sdk/core/core/test/cmocka/test_az_http.c [hunk 14]
```diff
@@ -357,8 +358,160 @@ ENABLE_PRECONDITION_CHECK_TESTS()
...
+static void test_http_response_header_validation_fail(void** state)
+{
+  (void)state;
+  {
+    az_http_response response = { 0 };
+    assert_return_code(
+        az_http_response_init(
+            &response,
+            AZ_SPAN_FROM_STR("HTTP/1.1 404 Not Found\r\n"
+                             "(Header11): Value11\r\n"
```
Comment (ahsonkhan) [inline sdk/core/core/test/cmocka/test_az_http.c:478]: Add test where the header name contains a white space (for example `header name: value`). I am not certain, but I think we'd want to reject that. Please double check.

File: sdk/core/core/src/az_http_response.c [hunk 15]
```diff
@@ -18,7 +19,21 @@ AZ_NODISCARD AZ_INLINE az_result _az_is_char(az_span slice, uint8_t c)
...
 }
 
-static AZ_NODISCARD az_result _az_is_a_colon(az_span slice) { return _az_is_char(slice, ':'); }
+static AZ_NODISCARD az_result _az_valid_header_name_to_colon(az_span slice)
+{
+  az_result is_colon_result = _az_is_char(slice, ':');
+  if (is_colon_result == AZ_OK)
+  {
+    return is_colon_result;
+  }
```
Comment (ahsonkhan) [inline sdk/core/core/src/az_http_response.c:28]: This has a lot of ceremony. Also, I think we end up allowing an empty header name and also access a pointer value out-of-bounds. I don't think that is valid. We should do the validation check before this.

A test like this behaves incorrectly with `az_http_response_get_next_header` returning `AZ_OK`:
```C
(void)state;
  {
    az_http_response response = { 0 };
    assert_return_code(
        az_http_response_init(
            &response,
            AZ_SPAN_FROM_STR("HTTP/1.1 404 Not Found\r\n"
                             ": Value11\r\n"
                             "\r\n"
                             "KKKKKJJJJJIIIIIHHHHHGGGGGFFFFFEEEEEDDDDDCCCCCBBBBBAAAAA")),
        AZ_OK);

    az_http_response_status_line status_line = { 0 };
    assert_return_code(az_http_response_get_status_line(&response, &status_line), AZ_OK);
    az_pair header = { 0 };
    az_result fail_header_result = az_http_response_get_next_header(&response, &header);
    assert_true(AZ_ERROR_HTTP_CORRUPT_RESPONSE_HEADER == fail_header_result);
  }
```

The `_az_span_scan_until` is doing a lot of unnecessary work, especially for what we are trying to do. We may want to get rid of and just use find.

Comment (ahsonkhan) [reply sdk/core/core/src/az_http_response.c:28]: `remaining` in `az_http_response_get_next_header` could be of size 0 here, and now we are reading from memory we shouldn't have access to:
https://github.com/Azure/azure-sdk-for-c/blob/de74386a0dc9051ba16229b705ad574fc1069a26/sdk/core/core/src/az_http_response.c#L171

File: sdk/core/core/inc/az_result.h [hunk 16]
```diff
@@ -128,6 +128,8 @@ typedef enum
   AZ_ERROR_HTTP_RESPONSE_OVERFLOW = _az_RESULT_MAKE_ERROR(_az_FACILITY_HTTP, 5),
   AZ_ERROR_HTTP_RESPONSE_COULDNT_RESOLVE_HOST = _az_RESULT_MAKE_ERROR(_az_FACILITY_HTTP, 6),
 
+  AZ_ERROR_HTTP_CORRUPT_RESPONSE_HEADER = _az_RESULT_MAKE_ERROR(_az_FACILITY_HTTP, 7),
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply sdk/core/core/inc/az_result.h:131]: got the name from @JeffreyRichter  :)

File: sdk/core/core/src/az_http_header_validation_private.h (hunk 12 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_http_header_validation_private.h:57]: header name can have spaces in the middle, like `this is my header:value`.  followinf RFC

File: sdk/core/core/src/az_http_header_validation_private.h (hunk 13 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_http_header_validation_private.h:115]: it's not capitalized on purpose so it can be use as toLower(uint8_t x).   

I won't expend more time adding tests for this edge case. It's already took more than it deserves.  We must wait until we are sure this is somehow worth it.

File: sdk/core/core/inc/az_result.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (JeffreyRichter) [reply sdk/core/core/inc/az_result.h:131]: Good point, let's use AZ_HTTP_RESPONSE_CORRUPT_HEADER.
Invalid sounds like it invalid in this context (but could be OK in a different context). Corrupt makes it sound like it bad in all contexts.

File: sdk/core/core/src/az_http_response.c (hunk 15 above)
Comment (vhvb1989) [reply sdk/core/core/src/az_http_response.c:28]: it can't be size 0 after reading a header.
RFC asks /r/n after all headers and before a body.
So, after reading a header that ends in `\r\n` we can have only two options
- Another header starts
or
- `\r\n` again as the end of headers

File: sdk/core/core/src/az_http_header_validation_private.h (hunk 13 above)
Comment (JeffreyRichter) [reply sdk/core/core/src/az_http_header_validation_private.h:115]: If you are doing some trick here, then you absolutely need comments to explain this.
But upper to lower is just add 0x20 and subtract to go the other way (for chars in the A-Z range)

File: sdk/core/core/src/az_http_header_validation_private.h (hunk 12 above)
Comment (ahsonkhan) [reply sdk/core/core/src/az_http_header_validation_private.h:57]: Can you point me to where it states that? I was looking at https://tools.ietf.org/html/rfc7230#section-3.2

File: sdk/core/core/src/az_http_response.c (hunk 15 above)
Comment (ahsonkhan) [reply sdk/core/core/src/az_http_response.c:28]: What is preventing the user providing incomplete or invalid response, or one that doesn't have any headers at all? It is just buffer with data that our API accepts and tries to parse. 

During that, we are reading memory that we shouldn't. That is not a good idea, in almost any circumstance.

Here's a test case:
```C

static void test_http_response_header_validation_fail(void** state)
{
  (void)state;
  {
    az_http_response response = { 0 };
    assert_return_code(
        az_http_response_init(
            &response,
            AZ_SPAN_FROM_STR("HTTP/1.1 404 Not Found\r\n")
            ),
        AZ_OK);

    az_http_response_status_line status_line = { 0 };
    assert_return_code(az_http_response_get_status_line(&response, &status_line), AZ_OK);
    az_pair header = { 0 };

    // This is reading from a memory address it shouldn't.
    az_result fail_header_result = az_http_response_get_next_header(&response, &header);
    assert_int_equal(AZ_ERROR_HTTP_CORRUPT_RESPONSE_HEADER, fail_header_result);
  }
}
```

File: sdk/core/core/src/az_http_header_validation_private.h (hunk 13 above)
Comment (ahsonkhan) [reply sdk/core/core/src/az_http_header_validation_private.h:115]: > it's not capitalized on purpose so it can be use as toLower(uint8_t x).

Where do you use it as `toLower` atm? I couldn't find it. Do you foresee us using that somewhere else, in the future? Maybe we address that concern once we find a need.

File: sdk/core/core/inc/az_result.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/core/core/inc/az_result.h:131]: That's fine by me. 

Corrupt makes it sound like in transition through the code, it got corrupted somehow (went from valid to invalid). But that's all semantics :)
